### PR TITLE
Adding initial writeup of wrapper-layer.md

### LINKED ISSUE #2: IDL approach
Complementary to #1.

In case it is useful, we could consider using an IDL to auto-generate libraries from a common core implementation.

Pros:
- No need to transpile.

Cons:
- Won't work for languages that don't use FFI.

### PR DESCRIPTION
**Pull request by sffc**
Also see #26

Closes #2

### CODE REVIEW

File: docs/wrapper-layer.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+A key piece of rolling out ergonomic OmnICU APIs for Rust and other target languages is the *wrapper layer*: code that sits between the client and the lower-level OmnICU libraries.  This document explains the design behind the OmnICU wrapper layer.
+
+## Background
+
+### Ergonomic API versus Logical API
+
+When a programmer writes code using Intl, the code they write should demonstrate their *intent*.  They should use what I'm calling the *ergonomic API* of OmnICU.
+
+A *logical API*, on the other hand, is one that is closely tied to specific to functional units, like currencies or measurement units.  A logical API function should require a specific hunk of data and perform a specific task.
```
Comment (nciric) [inline docs/wrapper-layer.md:12]: to specific functional units?

File: docs/wrapper-layer.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Although ergonomic APIs correspond with what users of i18n intend to do, from an implementation standpoint, they are often quite heavy-weight, pulling in many dependencies that may not be required for all call sites.  Intl.NumberFormat is an example: it supports currency formatting, which requires a lot of code and data, but when formatting a number that's not a currency, you shouldn't need to carry all that extra code and data.
+
+The goal of the *wrapper layer* is to provide an ergonomic API for clients, but map it to the logical API under the hood.
```
Comment (nciric) [inline docs/wrapper-layer.md:16]: So to clarify:
1. Wrapper layer implements ergonomic API
2. Is resolving user request (options, locale,...) into potential data fetch, and is passing data to logical API under the hood.

File: docs/wrapper-layer.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+### Host Language
+
+Since the ergonomic API depends on the host language (e.g., Objective-C calling into OmnICU via an FFI), the wrapper layer should be written in each individual host language, mapping to the logical API exported by the OmnICU library.
+
+This allows different languages to choose different styles for passing options into the library; for example, a Builder pattern could be used in Java, and an options bag could be used in JavaScript.
+
+### Code Slicing
+
+The wrapper layer should be suitable for tools to perform automatic dead code elimination.  By resolving options at compile time, the set of logical APIs required for a certain application can be known in time to perform static analysis; unused logical APIs can then be removed, reducing code and data size.
```
Comment (nciric) [inline docs/wrapper-layer.md:32]: This is for example critical to JS Closure that does all of the code elimination at compile time.

File: docs/wrapper-layer.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+Rust has great support for dead code elimination.  In languages where tooling for dead code elimination is not as strong, this should be taken into consideration when designing the ergonomic API specific to that language.  The ergonomic API might need to have a more logical flavor in order for code slicing to work as intended.
+
+### Data Loading
+
+In accordance with [principles.md](principles.md), the core library should not perform I/O; instead, data should be passed via a dynamic DataProvider ([data-pipeline.md](data-pipeline.md)).
+
+How does this fit into the wrapper layer model?
+
+1. The Ergonomic API receives a Data Provider.
```
Comment (nciric) [inline docs/wrapper-layer.md:44]: 1. This is not obvious from your diagram
2. Asking end client to always provide data provider in the API looks clunky bcs. it's mostly unrelated to functionality at hand @hagbard 

I'll open new issue to scope this better.

File: docs/wrapper-layer.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+2. The Wrapper Layer loads the data asynchronously based on the required functional units, and then passes the data to the Functional APIs.
+
+This model decouples the data provider from the core business logic; with calls to the data provider happening in the wrapper layer, the core library does not need to make the decision of sync vs. async, etc.
+
+### Limited Amount of Code
+
+One of the strengths of OmnICU is that the core business logic is written once and shared by many different environments.
+
+Since the wrapper layer needs to be implemented separately for each host language, as much business logic as possible should be moved into the code slicing layer.  The wrapper layer should transform from the ergonomic API to the logical API and load the required data for the logical API; nothing more.
```
Comment (nciric) [inline docs/wrapper-layer.md:53]: At this point maybe mention transpiling or IDL as potential ways to generate some of the binding logic?

File: docs/wrapper-layer.md (hunk 4 above)
Comment (nciric) [reply docs/wrapper-layer.md:44]: Added issue #30

File: docs/wrapper-layer.md (hunk 5 above)
Comment (sffc) [reply docs/wrapper-layer.md:53]: Added

File: docs/wrapper-layer.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
+The OmnICU Wrapper Layer
+========================
+
+A key piece of rolling out ergonomic OmnICU APIs for Rust and other target languages is the *wrapper layer*: code that sits between the client and the lower-level OmnICU libraries.  This document explains the design behind the OmnICU wrapper layer.
+
+## Background
+
+### Ergonomic API versus Logical API
+
+When a programmer writes code using Intl, the code they write should demonstrate their *intent*.  They should use what I'm calling the *ergonomic API* of OmnICU.
```
[the remark this answers is not in the document]
Comment (hagbard) [reply docs/wrapper-layer.md:10]: Just my $0.02: A good "ergonomic" API (I'd not head that term before, but I like it) would I think be one that makes the "business logic" of the caller more clearly aligned with the task being performed. It doesn't have to be "simple to use" (if the task if necessarily complex) but it should avoid adding more complexity than already exists in the problem at hand. You should try and align scopes of responsibility, lifecycle, ownership etc. (I think scope analysis is one of the keys to good API/code design). I'm writing up some notes on API design atm, so will share them when I have them (at some point I'll turn it into a talk).

[the remark this answers is not in the document]
Comment (sffc) [reply docs/wrapper-layer.md:10]: Added David's paragraph (slightly modified) to the doc.  Thanks!

File: docs/wrapper-layer.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sffc) [reply docs/wrapper-layer.md]: Thanks! Typo. Fixed

**Comment by sffc** [comment]
Ping for review/approval

File: docs/wrapper-layer.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (echeran) [reply docs/wrapper-layer.md]: Looks good to me.

**Comment by sffc** [comment]
Whoops, I should change `*ICU for X*` to `ICU4X`.  I'll do that in another PR right after this one.

**Pull request merged**